torchgeo/trainers/simclr.py

```diff
@@ -0,0 +1,198 @@
+import torch.nn as nn
+import torch.nn.functional as F
+from lightning.pytorch import LightningModule
+from torch import Tensor
+from torch.optim import AdamW, Optimizer
+from torch.optim.lr_scheduler import CosineAnnealingLR
+
+try:
+    from torch.optim.lr_scheduler import LRScheduler
+except ImportError:
+    from torch.optim.lr_scheduler import _LRScheduler as LRScheduler
+
+
+class SimCLRTask(LightningModule):  # type: ignore[misc]
+    """SimCLR: a simple framework for contrastive learning of visual representations.
+
+    Implementation based on:
+
+    * https://github.com/google-research/simclr
+    * https://lightning.ai/docs/pytorch/stable/notebooks/course_UvA-DL/13-contrastive-learning.html
+
+    If you use this trainer in your research, please cite the following papers:
+
+    * v1: https://arxiv.org/abs/2002.05709
+    * v2: https://arxiv.org/abs/2006.10029
+
+    .. versionadded:: 0.5
+    """  # noqa: E501
+
+    def __init__(
+        self,
+        model: str = "resnet50",
+        in_channels: int = 3,
+        version: int = 2,
+        layers: int = 3,
+        hidden_dim: int = 128,
+        lr: float = 4.8,
+        weight_decay: float = 1e-4,
+        max_epochs: int = 100,
+        temperature: float = 0.07,
+    ) -> None:
+        """Initialize a new SimCLRTask instance.
+
+        Args:
+            model: Name of the timm model to use.
+            in_channels: Number of input channels to model.
+            version: Version of SimCLR, 1--2.
+            layers: Number of layers in projection head (2 for v1 or 3+ for v2).
+            hidden_dim: Number of hidden dimensions in projection head.
+            lr: Learning rate (0.3 x batch_size / 256 is recommended).
+            weight_decay: Weight decay coefficient (1e-6 for v1 or 1e-4 for v2).
+            max_epochs: Maximum number of epochs to train for.
+            temperature: Temperature used in InfoNCE loss.
+        """
+        super().__init__()
+
+        assert version in range(1, 3)
+
+        self.save_hyperparameters()
+
+        self.model = timm.create_model(
+            model, in_chans=in_channels, num_classes=hidden_dim
+        )
+
+        # Projection head
+        # https://github.com/google-research/simclr/blob/master/model_util.py#L141
+        for i in range(layers):
+            if i == layers - 1:
+                # For the final layer, skip bias and ReLU
+                self.model.fc = nn.Sequential(
+                    self.model.fc, nn.Linear(hidden_dim, hidden_dim, bias=False)
+                )
+            else:
+                # For the middle layers, use bias and ReLU
+                self.model.fc = nn.Sequential(
+                    self.model.fc,
+                    nn.ReLU(inplace=True),
+                    nn.Linear(hidden_dim, hidden_dim, bias=True),
+                )
+
+        # TODO
+        # v1+: add global batch norm
+        # v2: add selective kernels, channel-wise attention mechanism, memory bank
```

Comment:
Not sure exactly how to make these changes, and I don't really want to change the architecture too much to ensure that our pre-trained weights can be loaded in a vanilla model. The memory bank only adds +1% performance, so I don't really think it's worth the complexity.

Reply:
I would say that when most papers use SimCLR, they use v1 without all the tricks that get the ~1% improvement. I think it would be better to keep it simple.

Reply:
The majority of the performance bump in v2 is thanks to the deeper projection head, which we have, so we should be good.

Reply:
Actually, in terms of performance bump:

* Bigger ResNets, SK, channel-wise attention: +29%
* Deeper projection head: +14%
* Memory bank: +1%

So memory bank isn't high on my priority list, but adding SK and channel-wise attention may be worth it.